### PR TITLE
Improve performance for WCSAxes imshow

### PR DESCRIPTION
**Pull request by thuiop**
### Description

This pull request aims to remove a performance issue when using the `imshow` of a `WCSAxes` instance. The issue is that this uses `minversion` to check the PIL version, which in the case of PIL uses the `packages_distributions` function which is extremely slow (in my case representing the vast majority of the time spent in the function). This PR avoids this by using `hasattr` on the module to assess which import should be used.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
This is fine in principle but it'd be preferable to stick with our local idiom for such cases: can you rewrite it by defining a module-level bool constant `PILLOW_LT_9_1` ?

**Comment by thuiop** [comment]
Hello, happy to do it. I would however like to note beforehand that this is a lazy import, and it is not use elsewhere in that file. I can move the conditionals at the module-level, but in that case it could be easier to directly set FLIP_TOP_TO_BOTTOM there rather than defining an extra intermediary constant?

**Comment by neutrinoceros** [comment]
Yes, that's probably best.

**Comment by neutrinoceros** [comment]
(Although note that the intermediary constant, `HAS_PIL`, should already be defined in `astropy.utils.compat.optional_deps`)

File: astropy/visualization/wcsaxes/core.py
Comment (neutrinoceros) [inline astropy/visualization/wcsaxes/core.py:31]: We should still use `minvesrion` here; it's much less of a hassle when it's not called within a loop.

**Review by neutrinoceros — CHANGES_REQUESTED** [review]

**Comment by eerovaher** [comment]
[`pillow` 9.1 was released in April 2022](https://pypi.org/project/pillow/9.1.0/), so [SPEC 0](https://scientific-python.org/specs/spec-0000/) allows us to drop support for older versions, in which case those version checks would not be needed at all.

**Comment by pllim** [comment]
Bumping minversion might overwhelm a first time contributor. If this is good enough to fix the immediate problem, maybe bumping minversion can be a follow-up PR by a maintainer.

Or if the author prefers a maintainer to take over here?

**Comment by pllim** [comment]
p.s. Do we add change log if we bump minversion of optional deps? I can't remember and no time to check right now.

**Comment by thuiop** [comment]
I am fine if you want to bump the version here, this would indeed be the cleanest way. I am happy if someone wants to take over the PR here to do it as I may miss something if I do it.

**Comment by neutrinoceros** [comment]
Sorry for the delay, I've been a little under the weather lately.
I just pushed a minimal fix to get CI + linters to pass again. In my opinion this is good to go as is, and bumping our minimal requirement for pillow is both out of scope and possibly more problematic than we'd like: pillow is actually not declared as an optional dependency yet; we only get it via `matplotlib`, which itself isn't a hard dependency, meaning its lower bound cannot strictly be enforced in practice: even though `astropy[recommended]` requires `matplotlib>=3.5`, one wouldn't get any sort of warning or error from the following sequence of commands
```shell
pip install astropy==6.1.5
pip install matplotlib==3.4.0
```

anyway, we can discuss that in another issue (though I don't think we need to), but it feels enough of a separate problem that it shouldn't block this PR. I'll merge when tests pass. Thanks @thuiop for going down this rabbit hole with us !

**Review by neutrinoceros — APPROVED** [review]

**Comment by thuiop** [comment]
Thanks for the quick and thorough handling! This was a low-hanging bottleneck which I ran into, I am happy it has been addressed.

**Pull request merged**
